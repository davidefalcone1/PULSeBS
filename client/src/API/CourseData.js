//usata per la retrieve della lista dei corsi di un docente
class CourseData{
    constructor(courseId, courseName, teacherId, 
        normalBookingsAvgWeek, cancelledBookingsAvgWeek, waitingBookingsAvgWeek, 
        normalBookingsAvgMonth, cancelledBookingsAvgMonth, waitingBookingsAvgMonth,
        attendanceCountAvgMonth, attendanceCountAvgWeek ) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
        this.normalBookingsAvgWeek = normalBookingsAvgWeek;
        this.cancelledBookingsAvgWeek = cancelledBookingsAvgWeek;
        this.waitingBookingsAvgWeek = waitingBookingsAvgWeek;
        this.normalBookingsAvgMonth = normalBookingsAvgMonth;
        this.cancelledBookingsAvgMonth = cancelledBookingsAvgMonth;
        this.waitingBookingsAvgMonth = waitingBookingsAvgMonth;
        this.attendanceCountAvgWeek = attendanceCountAvgWeek;
        this.attendanceCountAvgMonth = attendanceCountAvgMonth;
    }

    static fromJson(json){
        const temp =  Object.assign(new CourseData(), json);
        return temp;
    }
}

export default CourseData;