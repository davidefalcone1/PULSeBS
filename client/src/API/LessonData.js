import moment from 'moment';

class LessonData {
    constructor(scheduleId, courseId, startDate, endDate, occupiedSeats, availableSeats,
        isLessonCancelled, isLessonRemote, classroom
        /*, normalBookings, cancelledBookings, waitingBookings */) {
        if (scheduleId)
            this.scheduleId = scheduleId;
        this.courseId = courseId;
        this.startDate = moment(new Date(startDate));
        this.endDate = moment(new Date(endDate));
        this.occupiedSeats = occupiedSeats;
        this.availableSeats = availableSeats;
        this.isLessonCancelled = isLessonCancelled;
        this.isLessonRemote = isLessonRemote;
        this.classroom = classroom

        //TODO use the correct variables
        this.normalBookings = 0;
        this.cancelledBookings = 0;
        this.waitingBookings = 0;
    }

    static fromJson(json) {
        const temp = Object.assign(new LessonData(), json);
        temp.startDate = moment(new Date(temp.startDate));
        temp.endDate = moment(new Date(temp.endDate));
        return temp;
    }
}

export default LessonData;