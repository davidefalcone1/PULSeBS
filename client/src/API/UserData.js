class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

    static fromJson(json){
        const temp =  Object.assign(new UserData(), json);
        return temp;
    }
}

export default UserData;