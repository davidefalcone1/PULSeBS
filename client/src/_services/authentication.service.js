import { BehaviorSubject } from 'rxjs';


const currentUserSubject = new BehaviorSubject(JSON.parse(localStorage.getItem('user')));

export const authenticationService = {
    login,
    logout,
    user: currentUserSubject.asObservable(),
    get currentUserValue () { return currentUserSubject.value }
};

function handleResponse(response) {
    return response.text().then(text => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if ([401, 403].indexOf(response.status) !== -1) {
                // auto logout if 401 Unauthorized or 403 Forbidden response returned from api
                authenticationService.logout();
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }

        return data;
    });
}

function login(username, password) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ username, password })
    };
    return new Promise(async function (resolve, reject) {
        try {
            const response = await fetch('/users/authenticate', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({ username, password }),
            });
            const user = await response.json().then(handleResponse);
            if (response.ok) {
                localStorage.setItem('user', JSON.stringify(user));
                currentUserSubject.next(user);
                resolve(user);
            }
            else
                reject();
        }
        catch (e) {
            reject();
        }
    });

}
function logout() {

    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
    };
    localStorage.removeItem('user');
    currentUserSubject.next(null);
    return fetch('/logout',requestOptions);
}