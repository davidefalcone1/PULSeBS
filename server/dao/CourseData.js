//usata per la retrieve della lista dei corsi
class CourseData{
    constructor(courseId, courseName, teacherId) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
    }
}

module.exports = CourseData;