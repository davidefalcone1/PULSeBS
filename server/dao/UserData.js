class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

}

module.exports = UserData;