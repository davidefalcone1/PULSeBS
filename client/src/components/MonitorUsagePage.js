import React from 'react';
import {AuthContext} from '../_services/AuthContext';
import {Redirect} from 'react-router-dom';
import ListGroup from 'react-bootstrap/esm/ListGroup';
import Accordion from 'react-bootstrap/esm/Accordion';
import Card from 'react-bootstrap/esm/Card';
import Button from 'react-bootstrap/esm/Button';
import Badge from 'react-bootstrap/esm/Badge';
import Row from 'react-bootstrap/esm/Row';
import Col from 'react-bootstrap/esm/Col';

const monitorUsagePage = (props) => {
  return(
    <AuthContext.Consumer>
      {(context) => (
        <>        
          {context.user && 
            <>
              {props.courses && props.lessons &&
                <div style={{padding: "15px"}}>
                  <OverallStatistics courses = {props.courses}/>
                  <Accordion>
                    {props.courses.map((course) => //per ogni mio corso
                      <Card>
                        <Card.Header>
                            <Accordion.Toggle as={Button} variant="link" eventKey = {"course-" + course.courseName}>
                              <CourseHeader course = {course.courseName}/>
                            </Accordion.Toggle>
                        </Card.Header>
                        <Accordion.Collapse eventKey={"course-" + course.courseName}>
                          <Card.Body>
                            <CourseHeaderStatics course={course}/>
                            <Accordion>
                              {props.lessons.map((lesson) => //per ogni lezione del mio corso
                                (lesson.courseId === course.courseId) &&
                                  <Card>
                                    <Card.Header>
                                      <div className="d-flex w-100 pt-3 justify-content-between no-gutters">
                                        <Accordion.Toggle as={Button} variant="link" eventKey={course.courseName + "-" + lesson.scheduleId}>
                                          <LessonHeader id = {lesson.scheduleId} startDate = {lesson.startDate} endDate = {lesson.endDate}
                                            isLessonRemote={lesson.isLessonRemote} isLessonCancelled={lesson.isLessonCancelled}/>
                                        </Accordion.Toggle>
                                      </div>
                                    </Card.Header>
                                    <Accordion.Collapse eventKey={course.courseName + "-" + lesson.scheduleId}>
                                      <Card.Body>
                                        <LessonHeaderStatistics id = {lesson.scheduleId + "statistics"} normalBookings = {lesson.normalBookings}
                                          cancelledBookings = {lesson.cancelledBookings} waitingBookings = {lesson.waitingBookings} attendanceCount={lesson.attendanceCount}/>
                                      </Card.Body>
                                    </Accordion.Collapse>
                                  </Card>
                              )}
                            </Accordion>
                          </Card.Body>
                        </Accordion.Collapse>
                      </Card>
                    )}
                  </Accordion>
                </div>
                }
            </>}
          {!context.user && <Redirect to="/login"/>}
      </>
    )}
    </AuthContext.Consumer>
  );
}

function computeStatistics(courses){
  const result = {
    normalBookingsAvgWeek: 0,
    cancelledBookingsAvgWeek: 0,
    waitingBookingsAvgWeek: 0,
    normalBookingsAvgMonth: 0,
    cancelledBookingsAvgMonth: 0,
    waitingBookingsAvgMonth: 0,
    attendanceCountAvgWeek: 0,
    attendanceCountAvgMonth: 0
  };
  for(let course of courses){
    result.normalBookingsAvgMonth += course.normalBookingsAvgMonth;
    result.normalBookingsAvgWeek += course.normalBookingsAvgWeek;
    result.waitingBookingsAvgMonth += course.waitingBookingsAvgMonth;
    result.waitingBookingsAvgWeek += course.waitingBookingsAvgWeek;
    result.cancelledBookingsAvgMonth += course.cancelledBookingsAvgMonth;
    result.cancelledBookingsAvgWeek += course.cancelledBookingsAvgWeek;
    result.attendanceCountAvgMonth += course.attendanceCountAvgMonth;
    result.attendanceCountAvgWeek += course.attendanceCountAvgWeek;
  }
  result.normalBookingsAvgMonth /= courses.length;
  result.normalBookingsAvgWeek /= courses.length;
  result.waitingBookingsAvgMonth /= courses.length;
  result.waitingBookingsAvgWeek /= courses.length;
  result.cancelledBookingsAvgMonth /= courses.length;
  result.cancelledBookingsAvgWeek /= courses.length;
  result.attendanceCountAvgMonth /= courses.length;
  result.attendanceCountAvgWeek /= courses.length;
  return result;
}

function OverallStatistics(props) {
  const statistics = computeStatistics(props.courses);
  return (
    <div id={props.id}>
      <Row>
        <Col>
          <h5>Statistics per week</h5>
          <h6>
            {statistics.normalBookingsAvgWeek + statistics.cancelledBookingsAvgWeek + statistics.waitingBookingsAvgWeek} total bookings,{' '}
            {statistics.normalBookingsAvgWeek} actual bookings, {' '}
            {statistics.cancelledBookingsAvgWeek} cancelled bookings, {' '}
            {statistics.waitingBookingsAvgWeek} waiting bookings, {' '}
            {statistics.attendanceCountAvgWeek} attendance {' '}
          </h6>
        </Col>
        <Col>
          <h5>Statistics per month</h5>
          <h6>
            {statistics.normalBookingsAvgMonth + statistics.cancelledBookingsAvgMonth + statistics.waitingBookingsAvgMonth} total bookings, {' '}
            {statistics.normalBookingsAvgMonth} actual bookings, {' '}
            {statistics.cancelledBookingsAvgMonth} cancelled bookings, 
            {statistics.waitingBookingsAvgMonth} waiting bookings,
            {statistics.attendanceCountAvgMonth} attendance
          </h6>
        </Col>
      </Row>
      <br/>
    </div>
  );
}

function CourseHeader(props) {
  return (
    <div className="d-flex w-100 pt-3 justify-content-between no-gutters" id={"course-" + props.course}>
      <h4>{props.course}</h4>
    </div>
  );
}
function LessonHeader(props) {
  return (
    <>
      <div id={"lesson-" + props.startDate + "----" + props.endDate}>
        <h6>
          Lezione del {props.startDate.format("ddd DD-MM-YYYY HH:mm").toString()} -- {props.endDate.format("ddd DD-MM-YYYY HH:mm").toString()}
          {' '}
          {props.isLessonRemote && !props.isLessonCancelled && <Badge pill variant="warning">Remote</Badge>}
          {props.isLessonCancelled && <Badge pill variant="secondary">Cancelled</Badge>}
          {' '}
        </h6>
      </div>
    </>
  );
}

function LessonHeaderStatistics(props){
  debugger;
  return (
    <>
      <h6>{props.normalBookings} Bookings</h6>
      <h6>{props.cancelledBookings} Cancelled bookings</h6>
      <h6>{props.waitingBookings} Waiting bookings</h6>
      <h6>{props.normalBookings + props.cancelledBookings + props.waitingBookings} Total bookings</h6>
      <h6>{props.attendanceCount} Attendance</h6>
    </>
  );
}

function CourseHeaderStatics(props) {
  return (
    <div id={props.course.id}>
      <Row>
        <Col>
          <h5>Statistics per week</h5>
          <h6>{props.course.normalBookingsAvgWeek} Bookings</h6>
          <h6>{props.course.cancelledBookingsAvgWeek} Cancelled bookings</h6>
          <h6>{props.course.waitingBookingsAvgWeek} Waiting bookings</h6>
          <h6>{props.course.normalBookingsAvgWeek + props.course.cancelledBookingsAvgWeek + props.course.waitingBookingsAvgWeek} Total </h6>
          <h6>{props.course.attendanceCountAvgWeek} Attendance</h6>
        </Col>
        <Col>
          <h5>Statistics per month</h5>
          <h6>{props.course.normalBookingsAvgMonth} Bookings</h6>
          <h6>{props.course.cancelledBookingsAvgMonth} Cancelled bookings</h6>
          <h6>{props.course.waitingBookingsAvgMonth} Waiting bookings</h6>
          <h6>{props.course.normalBookingsAvgMonth + props.course.cancelledBookingsAvgMonth + props.course.waitingBookingsAvgMonth} Total </h6>
          <h6>{props.course.attendanceCountAvgMonth} Attendance</h6>
        </Col>
      </Row>
      <br/>
    </div>
  );
}

export default monitorUsagePage;
