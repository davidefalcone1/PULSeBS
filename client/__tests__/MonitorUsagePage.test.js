import React from 'react';
import renderer from 'react-test-renderer';
import { AuthContext } from '../src/_services/AuthContext';
import MonitorUsagePage from '../src/components/MonitorUsagePage';
import { BrowserRouter } from 'react-router-dom';
import UserData from '../src/API/UserData';
import CourseData from '../src/API/CourseData';
import LessonData from '../src/API/LessonData';
import moment from 'moment';

describe('MonitorUsagePage', ()=>{
    test('Renders correctly', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{user: new UserData(1, '12345', 'Davide Falcone', 'davide.falcone@polito.it')}}>
                    <MonitorUsagePage courses={[new CourseData('XY123', 'Software engineering 2', '654321', 2, 3, 5, 6, 7, 3, 5, 7)]} lessons={[new LessonData(1, 'XY123', moment('2020-09-08T14:30'), moment('2020-09-08T17:00'), 9, 90, false, true, 'A1', 100, 4, 33, 0)]}/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});