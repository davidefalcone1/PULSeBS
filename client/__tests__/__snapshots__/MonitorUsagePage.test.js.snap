// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitorUsagePage Renders correctly 1`] = `
<div
  style={
    Object {
      "padding": "15px",
    }
  }
>
  <div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <h5>
          Statistics per week (Average)
        </h5>
        <h6>
          10.00
           Total bookings
        </h6>
        <h6>
          2.00
           Actual bookings
        </h6>
        <h6>
          3.00
           Cancelled bookings
        </h6>
        <h6>
          5.00
           Waiting bookings
        </h6>
        <h6>
          7.00
           Attendance
        </h6>
      </div>
      <div
        className="col"
      >
        <h5>
          Statistics per month (Average)
        </h5>
        <h6>
          16.00
           Total bookings
        </h6>
        <h6>
          6.00
           Actual bookings
        </h6>
        <h6>
          7.00
           Cancelled bookings
        </h6>
        <h6>
          3.00
           Waiting bookings
        </h6>
        <h6>
          5.00
           Attendance
        </h6>
      </div>
    </div>
    <br />
  </div>
  <div
    className="accordion"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <div
            className="d-flex w-100 pt-3 justify-content-between no-gutters"
            id="course-Software engineering 2"
          >
            <h4>
              Software engineering 2
            </h4>
          </div>
        </button>
      </div>
      <div
        aria-expanded={null}
        className="collapse"
      >
        <div
          className="card-body"
        >
          <div>
            <div
              className="row"
            >
              <div
                className="col"
              >
                <h5>
                  Statistics per week (Average)
                </h5>
                <h6>
                  2
                   Bookings
                </h6>
                <h6>
                  3
                   Cancelled bookings
                </h6>
                <h6>
                  5
                   Waiting bookings
                </h6>
                <h6>
                  10
                   Total 
                </h6>
                <h6>
                  7
                   Attendance
                </h6>
              </div>
              <div
                className="col"
              >
                <h5>
                  Statistics per month (Average)
                </h5>
                <h6>
                  6
                   Bookings
                </h6>
                <h6>
                  7
                   Cancelled bookings
                </h6>
                <h6>
                  3
                   Waiting bookings
                </h6>
                <h6>
                  16
                   Total 
                </h6>
                <h6>
                  5
                   Attendance
                </h6>
              </div>
            </div>
            <br />
          </div>
          <div
            className="accordion"
          >
            <div
              className="card"
            >
              <div
                className="card-header"
              >
                <div
                  className="d-flex w-100 pt-3 justify-content-between no-gutters"
                >
                  <button
                    className="btn btn-link"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <div
                      id="lesson-1599568200000----1599577200000"
                    >
                      <h6>
                        Lezione del 
                        Tue 08-09-2020 14:30
                         -- 
                        Tue 08-09-2020 17:00
                         
                        <span
                          className="badge badge-pill badge-warning"
                        >
                          Remote
                        </span>
                         
                      </h6>
                    </div>
                  </button>
                </div>
              </div>
              <div
                aria-expanded={null}
                className="collapse"
              >
                <div
                  className="card-body"
                >
                  <h6>
                    100
                     Bookings
                  </h6>
                  <h6>
                    4
                     Cancelled bookings
                  </h6>
                  <h6>
                    33
                     Waiting bookings
                  </h6>
                  <h6>
                    137
                     Total bookings
                  </h6>
                  <h6>
                    0
                     Attendance
                  </h6>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
